File: linked_list.cpp
#include <cstdio>
#include <memory>
#include <utility>

struct Node {
    int value;
    std::unique_ptr<Node> next;
};

class LinkedList {
public:
    LinkedList() = default;
    LinkedList(const LinkedList&) = delete;
    LinkedList& operator=(const LinkedList&) = delete;

    void insert(int v) {
        auto n = std::make_unique<Node>();
        n->value = v;
        n->next = std::move(head_);
        head_ = std::move(n);
    }

    bool find(int v) const {
        const Node* cur = head_.get();
        while (cur) {
            if (cur->value == v) return true;
            cur = cur->next.get();
        }
        return false;
    }

    bool remove(int v) {
        std::unique_ptr<Node>* cur = &head_;
        while (*cur) {
            if ((*cur)->value == v) {
                *cur = std::move((*cur)->next);
                return true;
            }
            cur = &(*cur)->next;
        }
        return false;
    }

    ~LinkedList() {
        auto cur = std::move(head_);
        while (cur) {
            auto next = std::move(cur->next);
            cur = std::move(next);
        }
    }

    void print() const {
        std::printf("[");
        const Node* cur = head_.get();
        bool first = true;
        while (cur) {
            std::printf("%s%d", first ? "" : ", ", cur->value);
            first = false;
            cur = cur->next.get();
        }
        std::printf("]\n");
    }

private:
    std::unique_ptr<Node> head_;
};

int main() {
    LinkedList list;
    list.insert(3);
    list.insert(2);
    list.insert(1);
    std::printf("after inserts:  ");
    list.print();

    std::printf("find(2):        %s\n", list.find(2)    ? "true" : "false");
    std::printf("find(9):        %s\n", list.find(9)    ? "true" : "false");

    std::printf("delete(2):      %s\n", list.remove(2)  ? "true" : "false");
    std::printf("after delete:   ");
    list.print();

    std::printf("delete(1):      %s\n", list.remove(1)  ? "true" : "false");
    std::printf("after delete:   ");
    list.print();

    std::printf("delete(9):      %s\n", list.remove(9)  ? "true" : "false");
    std::printf("after delete:   ");
    list.print();
    return 0;
}

File: linked_list.rs
struct Node {
    value: i32,
    next: Option<Box<Node>>,
}

pub struct LinkedList {
    head: Option<Box<Node>>,
}

impl LinkedList {
    pub fn new() -> Self {
        LinkedList { head: None }
    }

    pub fn insert(&mut self, value: i32) {
        self.head = Some(Box::new(Node {
            value,
            next: self.head.take(),
        }));
    }

    pub fn find(&self, value: i32) -> bool {
        let mut current = &self.head;
        while let Some(node) = current {
            if node.value == value {
                return true;
            }
            current = &node.next;
        }
        false
    }

    pub fn delete(&mut self, value: i32) -> bool {
        let mut current = &mut self.head;
        loop {
            match current {
                None => return false,
                Some(node) if node.value == value => {
                    *current = node.next.take();
                    return true;
                }
                Some(node) => {
                    current = &mut node.next;
                }
            }
        }
    }
}

impl Drop for LinkedList {
    fn drop(&mut self) {
        let mut current = self.head.take();
        while let Some(mut node) = current {
            current = node.next.take();
        }
    }
}

impl std::fmt::Display for LinkedList {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "[")?;
        let mut current = &self.head;
        let mut first = true;
        while let Some(node) = current {
            if !first {
                write!(f, ", ")?;
            }
            write!(f, "{}", node.value)?;
            first = false;
            current = &node.next;
        }
        write!(f, "]")
    }
}

fn main() {
    let mut list = LinkedList::new();
    list.insert(3);
    list.insert(2);
    list.insert(1);
    println!("after inserts:  {}", list);

    println!("find(2):        {}", list.find(2));
    println!("find(9):        {}", list.find(9));

    println!("delete(2):      {}", list.delete(2));
    println!("after delete:   {}", list);

    println!("delete(1):      {}", list.delete(1));
    println!("after delete:   {}", list);

    println!("delete(9):      {}", list.delete(9));
    println!("after delete:   {}", list);
}
